Lib/functools.py

```diff
@@ -653,6 +653,48 @@ def cache(user_function, /):
     return lru_cache(maxsize=None)(user_function)
 
 
+################################################################################
+### Function variant registry
+################################################################################
+
+# {key: (max lineno, [variant])}
+_variant_registry = {}
```

Comment:
`collections` is already imported at the top of the file -- why not use a `defaultdict`?

Reply:
No strong reason; a plain dict works fine here and it avoids accidentally adding key-value pairs if we subscript it.

Reply:
Fair enough!

Another point's just occurred to me, however: what happens in this situation?

```python
@singledispatch
def func(arg: object) -> None: ...

@func.register
def _(arg: str) -> None: ...

del func
```

Will `_variant_registry` retain a reference to the variants of `func`, even after `func` has been `del`'d? That seems problematic. (The same point applies to the use with overloads.)

Maybe we should use a `weakref.WeakValueDictionary` instead?

Reply:
The problem is with overloads you have the undecorated implementation function, which is going to overwrite all the variants in the namespace - so we kinda have to keep them alive, unless either some new decorator is required or functions themselves do something.

Reply:
Yeah. weakrefs won't work with `@overload`. Which maybe brings us back to @rhettinger's idea of somehow hooking into the function-creation machinery and storing the overloads on the function object itself.

As a halfway house, we could maybe add a `weakref_only` optional parameter to `register_variant`, which means only a weakref of the variant is stored in the list instead of a strong reference. We could use that option for `singledispatch`, but not for `overload`.

Or alternatively, we could always only store weakrefs here in `functools._variant_registry`, and change the implementation of `typing.overload` so that it _also_ stores a reference to the overloads in `typing.py`'s global namespace.

Reply:
`singledispatch` already uses a `WeakKeyDictionary` in _its_ internal registry of functions, though, to fix this exact problem — so I don't think it's that arcane. And the whole point of exposing `register_function` as a public method is so that it can be used by third-party packages as well as the stdlib; and now we're saying that they can't use `register_function` unless they add a `__del__` method.

I feel like `typing.overload` is the odd one out here, as it's the only one that doesn't retain references to the variants, so it should therefore be `typing.overload` that's modified, rather than the other use cases. 

The fix could be as simple as this (in `typing.py`):

```python
_overloads = []

def overload(func):
    _overloads.append(func)
    # rest of the implementation of overload as it was before
```

Reply:
That causes a different issue - in cases where the overload's executed multiple times (say nested functions), this would cause those to effectively leak memory. 

Correctly dealing with all the situations seems intractable in the current syntax, perhaps we need to require decorating the implementation function in overloads with a new decorator.

Reply:
Coming back to this, it occurred to me that there's little reason to register variants in singledispatch anyway, since singledispatch functions already have a `.registry` attribute. Maybe it was a mistake to try to fit singledispatch and overload in the same framework, since they have different requirements and characteristics. Instead of calling register_variant() from singledispatch(), we can make get_variants() aware of singledispatch functions and retrieve the variants directly. That way we don't have to worry about weakrefs.

Reply:
But if the registry is only for overloads, shouldn't it live in typing.py? In this case I'd vote for solving a more specialized problem over a more general one, since we don't seem to have additional instances of the problem in mind.

Reply:
Yes, one concern about putting the registry in typing would be that users would have to unconditionally `import typing` and they may not want to.

Here are the use cases I'm aware of:
- Dynamic type checkers that want to see all overloads (my motivating use case). This PR directly allows that.
- pydoc should show all overloads (seems obviously useful, but needs more work on top of this PR). `pydoc` doesn't currently `import typing` but it probably could.
- inspect.signature should show overloads (@AlexWaygood alluded to this, but it would require significantly more work and I'm not sure it's worth it). Making `inspect` `import typing` would be problematic.

Reply:
`pydoc` uses `inspect.signature` internally to get information about a function's signature if a user calls `help()` on a function. Since both `inspect.signature()` and `help()` currently give bad results for multi-signature functions, it seems to me that the best way of fixing both together is to improve `inspect.signature`'s treatment of these functions. 

Completely agree that this would be complex, though, and would probably need discussion on python-ideas first.

Reply:
Yes. To be clear the reason it's complex is that currently `inspect.signature` returns a `Signature` object, and there's currently no way for it to represent overloads. So we'd need a whole new class to represent an overloaded signature.

Reply:
Why would it be a problem that `inspect` might need to `import typing`? `inspect` itself is a very large module with tons of dependencies, and if you use it you're giving up performance already (introspection in general is a slow process).